gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,58 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar, Union
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData, OutputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self) -> tf.Tensor:
+        """
+        Training loss for models that encapsulate their data.
+        """
+        return -self.maximum_a_posteriori_objective()
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
```

Comment:
i'm not so sure about this name. If I have some model `model`, I could easily think `model.training_loss_closure` is the closure itself, in part because it's a noun. I reckon `build_training_loss_closure` might be more appropriate

Reply:
I think it's sufficiently unambigous - training_loss is the function that returns the training loss as a tensor and training_loss_closure is the corresponding function that returns you a closure which returns the training loss as a tensor. I like that they have the same prefix as this makes code completion easy.

Reply:
we need it to be unambiguous for most users, and I estimate I'd be one of those who would find it unclear

Reply:
Added docstrings to clarify this. I would hope that the combination of return type signature, docstring, and example use in the notebooks makes this sufficiently clear for the level of users we are targetting with GPflow.